### PR TITLE
Add js tests for wiring connections

### PR DESCRIPTION
**Pull request by jpajuelo**

### CODE REVIEW

File: src/wirecloud/platform/static/js/wirecloud/wiring/Connection.js [hunk 1]
```diff
@@ -44,178 +44,145 @@
...
         },
+        target: {
+            type: "",
+            id: null,
+            endpoint: ""
+        }
+    };
 
-        members: {
+    // ==================================================================================
```
Comment (aarranz) [inline src/wirecloud/platform/static/js/wirecloud/wiring/Connection.js]: Breaks code style

File: src/wirecloud/platform/static/js/wirecloud/wiring/Connection.js [hunk 2]
```diff
@@ -44,178 +44,145 @@
...
+    // PUBLIC MEMBERS
+    // ==================================================================================
```
Comment (aarranz) [inline src/wirecloud/platform/static/js/wirecloud/wiring/Connection.js]: Breaks code style

File: src/js_tests/wirecloud/ConnectionSpec.js [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+                expect(connection.readonly).toBe(true);
+            });
+        });
+
+        describe("toJSON()", function () {
+
+            it("should parse to JSON", function () {
+                var connection = new ns.Connection(wiringEngine, sourceEndpoint, targetEndpoint);
+
+                expect(JSON.parse(JSON.stringify(connection))).toEqual({
```
Comment (aarranz) [inline src/js_tests/wirecloud/ConnectionSpec.js]: Seems an integration tests.... please, write a unit test.

File: src/js_tests/wirecloud/ConnectionSpec.js [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+        describe("detach()", function () {
+
+            it("should disconnect two endpoints properly", function () {
+                var connection = new ns.Connection(wiringEngine, sourceEndpoint, targetEndpoint);
+
+                var callback = jasmine.createSpy("callback");
+
+                connection.addEventListener("detach", callback);
+                connection.establish();
+                connection.detach();
```
Comment (aarranz) [inline src/js_tests/wirecloud/ConnectionSpec.js]: `detach` should support method chaining. Change this line to:

```
expect(connection.detach()).toBe(connection);
```

File: src/js_tests/wirecloud/ConnectionSpec.js [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+                expect(callback.calls.count()).toEqual(1);
+            });
+
+            it("should ignore if connection is not established", function () {
+                var connection = new ns.Connection(wiringEngine, sourceEndpoint, targetEndpoint);
+
+                var callback = jasmine.createSpy("callback");
+
+                connection.addEventListener("detach", callback);
+                connection.detach();
```
Comment (aarranz) [inline src/js_tests/wirecloud/ConnectionSpec.js]: `detach` should support method chaining. Change this line to:

```
expect(connection.detach()).toBe(connection);
```

File: src/js_tests/wirecloud/ConnectionSpec.js [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+                var connection = new ns.Connection(wiringEngine, sourceEndpoint, targetEndpoint);
+
+                var callback1 = jasmine.createSpy("callback1");
+                var callback2 = jasmine.createSpy("callback2");
+
+                connection.addEventListener("detach", callback1);
+                connection.addEventListener("remove", callback2);
+
+                connection.establish();
+                connection.remove();
```
Comment (aarranz) [inline src/js_tests/wirecloud/ConnectionSpec.js]: `remove` should support method chaining. Change this line to:

```
expect(connection.remove()).toBe(connection);
```

File: src/js_tests/wirecloud/ConnectionSpec.js [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+        describe("updateEndpoint(endpoint)", function () {
+
+            it("should change the missing source-endpoint", function () {
+                var missingSourceEndpoint = new ns.GhostSourceEndpoint(widgetModel, "missing");
+                var connection = new ns.Connection(wiringEngine, missingSourceEndpoint, targetEndpoint);
+
+                var callback = jasmine.createSpy("callback");
+
+                connection.addEventListener("establish", callback);
+                connection.updateEndpoint(sourceEndpoint);
```
Comment (aarranz) [inline src/js_tests/wirecloud/ConnectionSpec.js]: `updateEndpoint` should support method chaining. Change this line to:

```
expect(connection.updateEndpoint(sourceEndpoint)).toBe(connection);
```

File: src/js_tests/wirecloud/ConnectionSpec.js [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+                var missingTargetEndpoint = new ns.GhostTargetEndpoint(operatorModel, "missing");
+                var connection = new ns.Connection(wiringEngine, sourceEndpoint, missingTargetEndpoint);
+
+                var callback = jasmine.createSpy("callback");
+
+                connection.addEventListener("establish", callback);
+                connection.updateEndpoint(targetEndpoint);
+
+                expect(connection.established).toBe(true);
+                expect(connection.target).toEqual(targetEndpoint);
```
Comment (aarranz) [inline src/js_tests/wirecloud/ConnectionSpec.js]: `updateEndpoint` should support method chaining. Change this line to:

```
expect(connection.updateEndpoint(targetEndpoint)).toBe(connection);
```

File: src/js_tests/wirecloud/ConnectionSpec.js [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+
+        describe("establish()", function () {
+
+            it("should connect two endpoints properly", function () {
+                var connection = new ns.Connection(wiringEngine, sourceEndpoint, targetEndpoint);
+
+                var callback = jasmine.createSpy("callback");
+
+                connection.addEventListener("establish", callback);
+                connection.establish();
```
Comment (aarranz) [inline src/js_tests/wirecloud/ConnectionSpec.js]: `establish` should support method chaining. Change this line to:

```
expect(connection.establish()).toBe(connection);
```

File: src/js_tests/wirecloud/ConnectionSpec.js [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+            });
+
+            it("should ignore if connection is established", function () {
+                var connection = new ns.Connection(wiringEngine, sourceEndpoint, targetEndpoint);
+
+                var callback = jasmine.createSpy("callback");
+
+                connection.addEventListener("establish", callback);
+                connection.establish();
+                connection.establish();
```
Comment (aarranz) [inline src/js_tests/wirecloud/ConnectionSpec.js]: `establish` should support method chaining. Change this line to:

```
expect(connection.establish()).toBe(connection);
```

File: src/js_tests/wirecloud/ConnectionSpec.js [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+                    readonly: false,
+                    source: sourceEndpoint.toJSON(),
+                    target: targetEndpoint.toJSON()
+                });
+            });
+        });
+
+        describe("equals(value)", function () {
+
+            it("should be equal", function () {
```
Comment (aarranz) [inline src/js_tests/wirecloud/ConnectionSpec.js]: This method deserves more tests, for example, one for testing that the `equals` method returns `false` if the endpoints are different, another for testing it returns `true` if both connections are the same ....

File: src/js_tests/wirecloud/ConnectionSpec.js [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+            });
+
+            it("should ignore if one of their endpoints is missing", function () {
+                var missingTargetEndpoint = new ns.GhostTargetEndpoint(operatorModel, "missing");
+                var connection = new ns.Connection(wiringEngine, sourceEndpoint, missingTargetEndpoint);
+
+                var callback = jasmine.createSpy("callback");
+
+                connection.addEventListener("establish", callback);
+                connection.establish();
```
Comment (aarranz) [inline src/js_tests/wirecloud/ConnectionSpec.js]: `establish` should support method chaining. Change this line to:

```
expect(connection.establish()).toBe(connection);
```

File: src/wirecloud/platform/static/js/wirecloud/wiring/Connection.js [hunk 13]
```diff
@@ -153,7 +153,7 @@
 
                 if (endpoint instanceof Wirecloud.wiring.SourceEndpoint) {
                     this.source = endpoint;
-                } else if (endpoint instanceof Wirecloud.wiring.TargetEndpoint) {
+                } else {
```
Comment (aarranz) [inline src/wirecloud/platform/static/js/wirecloud/wiring/Connection.js]: Better provide a message there.

```javascript
} else /* if (endpoint instanceof Wirecloud.wiring.TargetEndpoint) */ {
```

File: src/wirecloud/platform/static/js/wirecloud/wiring/Connection.js [hunk 14]
```diff
@@ -144,7 +144,7 @@
 
             updateEndpoint: function updateEndpoint(endpoint) {
 
-                if (!(endpoint instanceof Wirecloud.wiring.Endpoint)) {
+                if (!(endpoint instanceof Wirecloud.wiring.SourceEndpoint || endpoint instanceof Wirecloud.wiring.TargetEndpoint)) {
```
Comment (aarranz) [inline src/wirecloud/platform/static/js/wirecloud/wiring/Connection.js]: This test is not coherent with the launched exception, also it does not make sense. Revert this change.

File: src/js_tests/wirecloud/ConnectionSpec.js [hunk 15]
```diff
@@ -0,0 +1,251 @@
...
+
+                connection.addEventListener("establish", callback);
+                connection.updateEndpoint(targetEndpoint);
+
+                expect(connection.established).toBe(true);
+                expect(connection.target).toEqual(targetEndpoint);
+                expect(callback.calls.count()).toEqual(1);
+            });
+
+            it("should throw error if endpoint is not valid", function () {
```
Comment (aarranz) [inline src/js_tests/wirecloud/ConnectionSpec.js]: Would be great to test more invalid values. Please create a describe section with a "throws a TypeError exception if endpoint is not valid" and create a test using `null`,  a number, a string and an object (as minimum).

File: src/js_tests/wirecloud/ConnectionSpec.js [hunk 16]
```diff
@@ -0,0 +1,251 @@
...
+                connection.addEventListener("detach", callback);
+                connection.detach();
+
+                expect(callback.calls.count()).toEqual(0);
+            });
+        });
+
+        describe("remove()", function () {
+
+            it("should remove after disconnecting endpoints", function () {
```
Comment (aarranz) [inline src/js_tests/wirecloud/ConnectionSpec.js]: "should disconnect the connection"

File: src/js_tests/wirecloud/ConnectionSpec.js [hunk 17]
```diff
@@ -0,0 +1,251 @@
...
+                var callback = jasmine.createSpy("callback");
+
+                connection.addEventListener("detach", callback);
+                connection.detach();
+
+                expect(callback.calls.count()).toEqual(0);
+            });
+        });
+
+        describe("remove()", function () {
```
Comment (aarranz) [inline src/js_tests/wirecloud/ConnectionSpec.js]: Missing tests removing a disconnected connection

File: src/js_tests/wirecloud/ConnectionSpec.js [hunk 18]
```diff
@@ -0,0 +1,251 @@
...
+
+                connection.addEventListener("establish", callback);
+                connection.establish();
+                connection.establish();
+
+                expect(connection.established).toBe(true);
+                expect(callback.calls.count()).toEqual(1);
+            });
+
+            it("should ignore if one of their endpoints is missing", function () {
```
Comment (aarranz) [inline src/js_tests/wirecloud/ConnectionSpec.js]: Split into two tests:

"should do nothing if the source is a ghost endpoint"
"should do nothing if the target is a ghost endpoint"

File: src/js_tests/wirecloud/ConnectionSpec.js [hunk 19]
```diff
@@ -0,0 +1,251 @@
...
+                var callback = jasmine.createSpy("callback");
+
+                connection.addEventListener("establish", callback);
+                connection.establish();
+
+                expect(connection.established).toBe(true);
+                expect(callback.calls.count()).toEqual(1);
+            });
+
+            it("should ignore if connection is established", function () {
```
Comment (aarranz) [inline src/js_tests/wirecloud/ConnectionSpec.js]: "should do nothing if the connection is already established"

File: src/js_tests/wirecloud/ConnectionSpec.js (hunk 12 above)
Comment (aarranz) [inline src/js_tests/wirecloud/ConnectionSpec.js]: `establish` should support method chaining. Change this line to:

```
expect(connection.establish()).toBe(connection);
```
